File: 프로그래머스/3/12927. 야근 지수/야근 지수.py
import heapq
def solution(n, works):
    if sum(works) <= n: return 0
    # 최대힙을 위해서 -w를 저장하기로
    arr = [-w for w in works]
    heapq.heapify(arr)
    while n > 0:
        job = heapq.heappop(arr)
        if job == 0: break
        job += 1
        n -= 1
        heapq.heappush(arr,job)
    ans = sum(w**2 for w in arr)
    return ans

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;
class Solution {
    public long solution(int n, int[] works) {
        if (Arrays.stream(works).sum() <= n) return 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int w : works) {
            pq.offer(-w);
        }
        while (n > 0) {
            int job = pq.poll();
            if (job == 0) break;
            job++;
            n--;
            pq.offer(job);
        }
        long ans = 0;
        for (int e : pq) {
            ans += e*e;
        }
        return ans;
    }
}